python/ray/serve/model_wrappers.py

```diff
@@ -57,6 +58,57 @@ def unpack(output_arr):
     return batched, unpack
 
 
+def collate_dict_array(
+    input_list: List[Dict[str, np.ndarray]]
+) -> Tuple[
+    Dict[str, np.ndarray],
+    Callable[[Dict[str, np.ndarray]], List[Dict[str, np.ndarray]]],
+]:
+    batch_size = len(input_list)
+
+    # Check all input has the same dict keys.
+    input_keys = [set(item.keys()) for item in input_list]
+    batch_has_same_keys = input_keys.count(input_keys[0]) == batch_size
+    if not batch_has_same_keys:
+        raise ValueError(
+            f"The input batch contains dictionary of different keys: {input_keys}"
+        )
+
+    # Turn list[dict[str, array]] to dict[str, List[array]]
+    key_to_list = defaultdict(list)
+    for single_dict in input_list:
+        for key, arr in single_dict.items():
+            key_to_list[key].append(arr)
+
+    # Turn dict[str, List[array]] to dict[str, array]
+    batched_dict = {}
+    unpack_dict = {}
+    for key, list_of_arr in key_to_list.items():
+        arr, unpack_func = collate_array(list_of_arr)
+        batched_dict[key] = arr
+        unpack_dict[key] = unpack_func
+
+    def unpack(output_dict: Dict[str, np.ndarray]):
+        # short circuit behavior, assume users know what they are doing.
+        if isinstance(output_dict, list):
```

Comment:
When would a user pass in a list for the `output_dict`?

Reply:
This is part of DataBatchType design like {"image": tensor, "text": tensor}.

Reply:
Could we expand on the comment above this line? It's not obvious what `"assume users know what they are doing"` means from a first glance.

Reply:
sure!